### PR TITLE
fix(datepicker): reduce flatpickr reset on changes

### PR DESCRIPTION
**Pull request by yinglejia**
Closes IBM/carbon-components-angular#

Currently, the date picker component resets the flatpickr instance on all input changes. This is unnecessary for input changes irrelevant to flatpickr, and can cause problems in some cases esp. for range date picker. For example, if the caller updates the validation status when the user selects a start date, then the flatpickr instance will be reset (destroyed) before the user can select an end date.

This PR limits the reset to relevant input changes only.

#### Changelog

**New**

* {{new thing}}

**Changed**

* {{change thing}}

**Removed**

* {{removed thing}}

### CODE REVIEW

File: src/datepicker/datepicker.component.ts [hunk 1]
```diff
@@ -298,13 +298,27 @@ export class DatePicker implements
...
-				this.resetFlackpickrInstance();
+				this.resetFlatpickrInstance();
 			});
 		}
 	}
 
 	ngOnChanges(changes: SimpleChanges) {
-		this.resetFlackpickrInstance(changes.value);
+		// Reset the flatpickr instance on input changes that affect flatpickr.
+		const resetOnInputs = [
```
Comment (Donisius) [inline src/datepicker/datepicker.component.ts:308]:
```suggestion
		const flatpickerChangeKeys = [
```

File: src/datepicker/datepicker.component.ts [hunk 2]
```diff
@@ -298,13 +298,27 @@ export class DatePicker implements
...
+			"dateFormat",
+			"language",
+			"id",
+			"value",
+			"theme",
+			"plugins",
+			"flatpickrOptions"
+		];
+		const changeKeys = Object.keys(changes);
+		if (changeKeys.some(key => resetOnInputs.includes(key))) {
```
Comment (Donisius) [inline src/datepicker/datepicker.component.ts:319]:
```suggestion
		if (changeKeys.some(key => flatpickerChangeKeys.includes(key))) {
```

**Review by Donisius — CHANGES_REQUESTED** [review]
Thanks for doing this @yinglejia! Looks good overall just a few nitpicks

File: src/datepicker/datepicker.component.ts [hunk 3]
```diff
@@ -298,13 +298,27 @@ export class DatePicker implements
...
 	ngOnChanges(changes: SimpleChanges) {
-		this.resetFlackpickrInstance(changes.value);
+		// Reset the flatpickr instance on input changes that affect flatpickr.
+		const resetOnInputs = [
+			"range",
+			"dateFormat",
+			"language",
+			"id",
+			"value",
+			"theme",
```
[the remark this answers is not in the document]
Comment (yinglejia) [reply src/datepicker/datepicker.component.ts]: Seems not needed, so removed.

**Review by Donisius — APPROVED** [review]

**Review by zvonimirfras — APPROVED** [review]

**Pull request merged**